public/app/plugins/datasource/cloud-monitoring/datasource.test.ts

```diff
@@ -0,0 +1,42 @@
+import { get, set } from 'lodash';
+
+import * as ts from 'app/features/templating/template_srv';
+
+import { createMockQuery } from './__mocks__/cloudMonitoringQuery';
+import { createMockInstanceSetttings } from './__mocks__/instanceSettings';
+import { createTemplateVariables } from './__mocks__/utils';
+import Datasource from './datasource';
+
+const templateSrv = new ts.TemplateSrv();
+
+describe('Cloud Monitoring Datasource', () => {
+  describe('interpolateVariablesInQueries', () => {
+    beforeEach(() => {
+      templateSrv.init([]);
+    });
+
+    it('should leave a query unchanged if there are no template variables', () => {
+      const mockInstanceSettings = createMockInstanceSetttings();
+      const ds = new Datasource(mockInstanceSettings);
+      const query = createMockQuery({ hide: false });
+      const templateVariablesApplied = ds.interpolateVariablesInQueries([query], {});
+      expect(templateVariablesApplied[0]).toEqual(query);
+    });
+
+    it('should correctly apply template variables', () => {
+      const templateVariables = createTemplateVariables(['metricQuery.projectName']);
```

Comment:
I believe this util function is hurting more than helping, it's a bit difficult to know what's happening. You are passing a path as a string, which should match a real object path to work and generates something a bit obscure (you are losing all the benefits of typings). It also relies a lot in the implementation of templateSrv, which we don't need to care.

I believe you can simplify all this test with something like:

```
    it('should correctly apply template variables', () => {
      templateSrv.replace = jest.fn().mockReturnValue('myvar');
      const query = createMockQuery({ hide: false });
      const templatedQuery = ctx.ds.interpolateVariablesInQueries([query], {});
      expect(templatedQuery[0].metricQuery.projectName).toEqual('myvar');
    });
```

and test the same thing. You can even check if `.replace` was called with the right arguments if you want to.

Also, I believe this test is not testing anything new? This test should pass with teh current main so you are not testing the bug fix?

Reply:
This test wouldn't pass on current main because of the `hide: false`. The main change in this PR is that we now spread the whole query when applying template variables to ensure none of the original query properties are lost (which is why the bug appeared).

Your implementation for the test does make more sense, I'll make some changes on this 😊

Reply:
aah, gotcha, can you add that to the test description or something? (also a test with `hide: true` to see the difference?)